### PR TITLE
Fix/atomics

### PR DESCRIPTION
**Pull request by reazulhoque**
This PR will add native floating-point atomics for GPUs. This work will require llvm-spirv translator that supports the `SPV_EXT_shader_atomic_float_add` extension to generate the native floating point atomic instruction.

### CODE REVIEW

File: numba_dppy/extended_numba_itanium_mangler.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from numba.core import itanium_mangler, types
+from numba_dppy import target
+
+EXQ2ABI = {
```
Comment (diptorupd) [inline numba_dppy/extended_numba_itanium_mangler.py:18]: Did you not remove tjis code?

File: numba_dppy/ocl/atomics/atomic_helper.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    An enumeration of SYCL memory order.
+    Please refer:
+    https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/ONEAPI/atomic_enums.hpp#L25
+
+    =====================   ============
+    Memory Order            Enum value
+    =====================   ============
+    relaxed                 0
+    acquire                 1
+    __consume_unsupported   2
```
Comment (diptorupd) [inline numba_dppy/ocl/atomics/atomic_helper.py:29]: do we need this entry?

**Comment by diptorupd** [comment]
I will review this in detail later. One note for all the enums, please document them in detail. The links to Dpcpp source is not sufficient, moreover the links are subject to change if Dpcpp code changes.

File: numba_dppy/extended_numba_itanium_mangler.py (hunk 1 above)
Comment (reazulhoque) [reply numba_dppy/extended_numba_itanium_mangler.py]: I did, merged that branch, should be gone.

File: numba_dppy/config.py [hunk 3]
```diff
@@ -61,3 +61,8 @@ def _readenv(): ...
 OFFLOAD_DIAGNOSTICS = _readenv("NUMBA_DPPY_OFFLOAD_DIAGNOSTICS", int, 0)
 
 FALLBACK_ON_CPU = _readenv("NUMBA_DPPY_FALLBACK_ON_CPU", int, 1)
+
+# Activate Native floating point atomcis support for supported devices.
+# Requires llvm-spirv supporting the FP atomics extensio
+NATIVE_FP_ATOMICS = _readenv("NUMBA_DPPY_ACTIVATE_NATIVE_FP_ATOMICS", int, 0)
```
Comment (PokhodenkoSA) [inline numba_dppy/config.py:67]: It is better to place more general part of name first: ATOMICS_USE_NATIVE_FP

File: numba_dppy/ocl/oclimpl.py [hunk 4]
```diff
@@ -29,6 +29,9 @@
 from numba_dppy import target
 from . import stubs
 from numba_dppy.codegen import SPIR_DATA_LAYOUT
+from numba_dppy.ocl.atomics import atomic_helper
+import dpctl
+from numba_dppy.config import NATIVE_FP_ATOMICS
```
Comment (PokhodenkoSA) [inline numba_dppy/ocl/oclimpl.py:34]: Use isort if possible for sorting imports.

File: numba_dppy/target.py [hunk 5]
```diff
@@ -68,6 +68,7 @@ def load_additional_registries(self):
 
 
 LINK_ATOMIC = 111
+LLVM_SPIRV_ARGS = 112
```
Comment (PokhodenkoSA) [inline numba_dppy/target.py:71]: Magic numbers?

File: numba_dppy/spirv_generator.py [hunk 6]
```diff
@@ -77,17 +77,43 @@ def optimize(self, ipath, opath):
...
 
         check_call(["opt", opt_level_option, "-o", ipath + ".bc", ipath])
-        check_call(["llvm-spirv", "-o", opath, ipath + ".bc"])
+
+        if dppy_config.NATIVE_FP_ATOMICS == 1:
+            llvm_spirv_root = dppy_config.LLVM_SPIRV_ROOT
+
+            if llvm_spirv_root == "":
+                # try to find ONEAPI root
+                possible_oneapi_roots = ["/opt/intel/oneapi", "$A21_SDK_ROOT"]
```
Comment (diptorupd) [inline numba_dppy/spirv_generator.py:95]: We cannot have these in our code-base. Users should set it themselves using envars.

File: numba_dppy/config.py [hunk 7]
```diff
@@ -62,5 +62,9 @@ def _readenv(): ...
 
 FALLBACK_ON_CPU = _readenv("NUMBA_DPPY_FALLBACK_ON_CPU", int, 1)
 
+# Activate Native floating point atomcis support for supported devices.
+# Requires llvm-spirv supporting the FP atomics extensio
+NATIVE_FP_ATOMICS = _readenv("NUMBA_DPPY_ACTIVATE_ATOMCIS_FP_NATIVE", int, 0)
```
Comment (PokhodenkoSA) [inline numba_dppy/config.py:67]: Please, make the same name in both places.

File: numba_dppy/ocl/atomics/atomic_helper.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+
+class sycl_memory_order(Enum):
+    """
+    An enumeration of SYCL memory order. For more details please refer to
+    SYCL 2020 specification, section 3.8.3.1
+    (https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_memory_ordering).
+
+    For DPCPP implementation please refer:
+    https://github.com/intel/llvm/blob/sycl-nightly/20210507/sycl/include/CL/sycl/ONEAPI/atomic_enums.hpp#L25
```
Comment (PokhodenkoSA) [inline numba_dppy/ocl/atomics/atomic_helper.py:25]: We need protection for eliminate mismatch between DPCPP and our copy of enums.

File: numba_dppy/ocl/atomics/atomic_ops.cl [hunk 9]
```diff
@@ -4,47 +4,6 @@
 */
 #ifdef cl_khr_int64_base_atomics
   #pragma OPENCL EXTENSION cl_khr_int64_base_atomics: enable
-
-  long numba_dppy_atomic_add_i64_local(volatile __generic long *p, long val) {
-      long found = *p;
-      long expected;
-      do {
```
Comment (PokhodenkoSA) [inline numba_dppy/ocl/atomics/atomic_ops.cl:11]: Removed all integer functions. Only floats needed.

File: numba_dppy/config.py (hunk 7 above)
Comment (PokhodenkoSA) [reply numba_dppy/config.py:67]: Cython to use for auto update.

**Review by PokhodenkoSA — APPROVED** [review]

**Pull request merged**